File: scripts/purge-junk.mjs
/**
 * purge-junk.mjs — Step 1
 * Remove spam listings (20+ rows) and non-clothing junk from all cache rows.
 * Also checks a sample of URLs for dead listings.
 */
import postgres from "postgres";

const DATABASE_URL = "postgresql://postgres.cdjuosvljudidvyxdfwn:RJkU3AvtaV2BuBGy@aws-1-us-east-1.pooler.supabase.com:5432/postgres";
const client = postgres(DATABASE_URL, { ssl: { rejectUnauthorized: false }, max: 5, prepare: false });

const JUNK_TITLE = [
  "trading card","pokemon card","yugioh","yu-gi-oh","magic the gathering","mtg ",
  "sports card","collectible","funko","action figure","figurine",
  "video game","console","phone case","magnet frame","picture frame",
  "poster","sticker","art print","wall art","candle","mug","cup",
  "pillow","blanket","book","magazine","vinyl record"," dvd",
  "costume jewelry","jewelry set","piece new costume","eye candy",
  "padded no","gathering mtg","nwt victorias secret padded",
  "14 piece new","magnet frames set",
];

function isJunk(l) {
  const t = (l.title || "").toLowerCase();
  return JUNK_TITLE.some(s => t.includes(s));
}

async function checkDead(urls, n = 150) {
  const sample = [...urls].sort(() => Math.random() - 0.5).slice(0, n);
  const dead = new Set();
  const BATCH = 15;
  for (let i = 0; i < sample.length; i += BATCH) {
    const slice = sample.slice(i, i + BATCH);
    await Promise.all(slice.map(async url => {
      if (!url?.startsWith("https://www.depop.com/products/")) return;
      try {
        const res = await fetch(url, {
          method: "HEAD", redirect: "follow",
          signal: AbortSignal.timeout(5000),
          headers: { "User-Agent": "Mozilla/5.0" },
        });
        if (res.status === 404 || (res.url || "").includes("/search/")) dead.add(url);
      } catch { /* timeout = assume alive */ }
    }));
    process.stdout.write(`  URL check ${Math.min(i+BATCH, sample.length)}/${sample.length}\r`);
  }
  return dead;
}

async function run() {
  console.log("Finding spam URLs (20+ rows)…");
  const spamRows = await client`
    SELECT l->>'url' as url FROM depop_cache, jsonb_array_elements(listings) as l
    GROUP BY url HAVING COUNT(*) > 20
  `;
  const spamUrls = new Set(spamRows.map(r => r.url).filter(Boolean));
  console.log(`  ${spamUrls.size} spam URLs`);

  console.log("Loading all rows…");
  const rows = await client`SELECT query, listings FROM depop_cache`;

  // Collect unique product URLs for dead-check
  const allUrls = new Set();
  for (const row of rows)
    for (const l of (row.listings || []))
      if (l.url?.startsWith("https://www.depop.com/products/")) allUrls.add(l.url);

  console.log(`Spot-checking 150 URLs for dead listings…`);
  const deadUrls = await checkDead(allUrls);
  console.log(`\n  Dead: ${deadUrls.size}`);

  console.log("Purging…");
  let removed = 0;
  const BATCH = 20;
  for (let i = 0; i < rows.length; i += BATCH) {
    const slice = rows.slice(i, i + BATCH);
    await Promise.all(slice.map(row => {
      const before = row.listings.length;
      const cleaned = (row.listings || []).filter(l =>
        !spamUrls.has(l.url) && !deadUrls.has(l.url) && !isJunk(l)
      );
      removed += before - cleaned.length;
      const jsonStr = JSON.stringify(cleaned).replace(/'/g, "''");
      const q = row.query.replace(/'/g, "''");
      return client.unsafe(`UPDATE depop_cache SET listings = '${jsonStr}'::jsonb WHERE query = '${q}'`);
    }));
    if (i % 500 === 0) console.log(`  ${i}/${rows.length}…`);
  }
  console.log(`Done! Removed ${removed} junk/dead listings.`);
  await client.end();
}
run().catch(e => { console.error(e); process.exit(1); });

File: scripts/python/purge_junk.py
"""
purge_junk.py — Remove non-clothing and spam listings from the Depop cache.

HOW IT WORKS:
  Sometimes Depop search results include non-clothing items like trading cards,
  phone cases, posters, etc. This script scans every listing and removes anything
  that clearly isn't clothing based on keywords in the title.

  It also removes "spam" listings — items that appear in 20+ cache rows,
  which usually means they're generic/unrelated products that show up in every search.

WHEN TO RUN THIS:
  - After seeding new listings (to clean up any junk that got in)
  - If you notice weird non-clothing items appearing in the app

USAGE:
  pip install psycopg2-binary
  python scripts/python/purge_junk.py           # dry run — shows what would be removed
  python scripts/python/purge_junk.py --delete  # actually removes junk
"""

import sys
import json
import psycopg2
import psycopg2.extras

# ── CONFIG ────────────────────────────────────────────────────────────────────
DB_URL   = "postgresql://postgres.cdjuosvljudidvyxdfwn:RJkU3AvtaV2BuBGy@aws-1-us-east-1.pooler.supabase.com:5432/postgres"
DRY_RUN  = "--delete" not in sys.argv
SPAM_THRESHOLD = 20  # listings appearing in this many rows are considered spam

# ── JUNK KEYWORDS ─────────────────────────────────────────────────────────────
# If any of these phrases appear in a listing title, it gets removed.
# Add more if you see junk slipping through.
JUNK_KEYWORDS = [
    # Trading cards / collectibles
    "trading card", "pokemon card", "yugioh", "yu-gi-oh", "magic the gathering",
    "mtg ", "sports card", "collectible", "funko", "action figure", "figurine",

    # Electronics / misc
    "video game", "console", "phone case", "phone cover",

    # Home goods
    "picture frame", "magnet frame", "poster", "sticker", "art print",
    "wall art", "candle", "mug", "cup", "pillow", "blanket",

    # Books / media
    "book", "magazine", "vinyl record", " dvd", " cd ",

    # Jewelry (non-clothing)
    "costume jewelry", "jewelry set", "earring set", "necklace set",

    # Junk phrases that sneak through
    "piece new costume", "eye candy", "padded no", "gathering mtg",
    "nwt victorias secret padded", "14 piece new", "magnet frames set",
]


# ── JUNK DETECTOR ─────────────────────────────────────────────────────────────
def is_junk(listing):
    """Returns True if this listing is clearly not clothing."""
    title = (listing.get("title") or "").lower()
    return any(keyword in title for keyword in JUNK_KEYWORDS)


# ── DATABASE ──────────────────────────────────────────────────────────────────
def get_connection():
    return psycopg2.connect(DB_URL, sslmode="require")


def find_spam_urls(cursor):
    """
    Finds URLs that appear in too many cache rows.
    These are likely irrelevant items that show up in every search.
    """
    cursor.execute("""
        SELECT l->>'url' as url, COUNT(*) as cnt
        FROM depop_cache, jsonb_array_elements(listings) as l
        WHERE jsonb_typeof(listings) = 'array'
        GROUP BY url
        HAVING COUNT(*) > %s
    """, (SPAM_THRESHOLD,))
    rows = cursor.fetchall()
    spam = {row[0] for row in rows if row[0]}
    return spam


# ── MAIN ──────────────────────────────────────────────────────────────────────
def main():
    mode = "DRY RUN" if DRY_RUN else "LIVE — removing junk"
    print(f"\n🪡  Stitch Junk Purger ({mode})\n")

    conn   = get_connection()
    cursor = conn.cursor()

    # Step 1: Find spam URLs
    print("🔍  Finding spam URLs (appearing in too many rows)...")
    spam_urls = find_spam_urls(cursor)
    print(f"   Found {len(spam_urls)} spam URLs\n")

    # Step 2: Load all rows
    print("📦  Loading all cache rows...")
    cursor.execute("""
        SELECT query, listings FROM depop_cache
        WHERE listings IS NOT NULL AND listings::text NOT IN ('[]', 'null', '')
    """)
    rows = cursor.fetchall()
    print(f"   {len(rows)} rows loaded\n")

    # Step 3: Scan and clean
    total_removed = 0
    rows_updated  = 0
    rows_deleted  = 0

    for query, listings_raw in rows:
        # Parse listings
        if isinstance(listings_raw, str):
            try:
                listings = json.loads(listings_raw)
            except Exception:
                continue
        else:
            listings = listings_raw

        if not isinstance(listings, list) or not listings:
            continue

        # Filter out junk and spam
        cleaned = [
            l for l in listings
            if not is_junk(l) and l.get("url") not in spam_urls
        ]
        removed = len(listings) - len(cleaned)

        if removed == 0:
            continue  # nothing to remove

        total_removed += removed
        print(f"  🗑️  \"{query}\": removing {removed} junk listings ({len(listings)} → {len(cleaned)})")
        for l in listings:
            if is_junk(l):
                print(f"       junk: {l.get('title', '?')}")
            elif l.get("url") in spam_urls:
                print(f"       spam: {l.get('title', '?')}")

        if not DRY_RUN:
            if not cleaned:
                cursor.execute("DELETE FROM depop_cache WHERE query = %s", (query,))
                rows_deleted += 1
            else:
                cursor.execute(
                    "UPDATE depop_cache SET listings = %s::jsonb WHERE query = %s",
                    (json.dumps(cleaned), query)
                )
                rows_updated += 1
            conn.commit()

    cursor.close()
    conn.close()

    print(f"\n✅  Done!")
    print(f"   Junk removed: {total_removed} listings")
    if not DRY_RUN:
        print(f"   Rows updated: {rows_updated}")
        print(f"   Rows deleted: {rows_deleted}")
    else:
        print(f"\n   ℹ️  Dry run — re-run with --delete to actually remove them.")
    print()


if __name__ == "__main__":
    main()
